File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/forbid-empty-catch-block.ts
#!/usr/bin/env bun
import { SyntaxKind, createSourceFile, forEachChild, isCatchClause, type Node, type SourceFile } from "typescript@6.0.3";
import type { Finding, HarnessCheckRule, HarnessManifest, RuleContext } from "../harness-check-rule";

/**
 * Forbid empty catch blocks.
 */
export class ForbidEmptyCatchBlockRule implements HarnessCheckRule {
  static readonly category = "forbidEmptyCatchBlock";

  constructor(private readonly ctx: RuleContext) {}

  applies(_manifest: HarnessManifest): boolean {
    return true;
  }

  validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
    const sources = this.ctx.stackSources(manifest, "typescript");
    return sources.flatMap((file) => {
      const text = this.ctx.read(file);
      if (!text) {
        return [];
      }

      let sourceFile: SourceFile;
      try {
        sourceFile = createSourceFile(file, text, SyntaxKind.LatestVersion, true);
      } catch {
        return [
          {
            severity: this.ctx.severityOf(manifest, ForbidEmptyCatchBlockRule.category),
            category: ForbidEmptyCatchBlockRule.category,
            message: `failed to parse TypeScript: ${file}`,
          },
        ];
      }

      const findings: Finding[] = [];

      const visit = (node: Node): void => {
        if (isCatchClause(node)) {
          if (node.block.statements.length === 0) {
            const { line } = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile));
            findings.push({
              severity: this.ctx.severityOf(manifest, ForbidEmptyCatchBlockRule.category),
              category: ForbidEmptyCatchBlockRule.category,
              message: `${file}:${line + 1}: empty catch block; handle, rethrow, or convert to a Finding`,
            });
          }
        }
        forEachChild(node, visit);
      };

      visit(sourceFile);
      return findings;
    });
  }
}

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/forbid_empty_catch_block.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Forbid empty catch block rule."""

from collections.abc import Iterable

from pathlib import Path

import libcst as cst

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import parse_python, relative, severity_for, stack_sources

class ForbidEmptyCatchBlockRule(HarnessCheckRule):
    """Validate forbidEmptyCatchBlock check."""

    category = "forbidEmptyCatchBlock"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate forbidEmptyCatchBlock check."""
        severity = severity_for(manifest, self.category)
        sources = stack_sources(root, manifest, self.category)
        class _EmptyCatchFinder(cst.CSTVisitor):
            def __init__(self, rel_path: str) -> None:
                super().__init__()
                self.findings: list[Finding] = []
                self.rel_path = rel_path
            def visit_Try(self, node: cst.Try) -> bool:
                for handler in node.handlers:
                    if not isinstance(handler.body, cst.IndentedBlock):
                        continue
                    body_stmts = handler.body.body
                    if not body_stmts:
                        pos = self.get_metadata(cst.metadata.PositionProvider, handler)
                        self.findings.append(Finding(
                            severity,
                            self.category,
                            f"{self.rel_path}:{pos.start.line}: empty catch block; handle, rethrow, or convert to a Finding",
                        ))
                return True
        result = []
        for path in sources:
            tree, error = parse_python(path)
            if error is not None:
                result.append(Finding(
                    severity,
                    self.category,
                    f"{relative(path)}: syntax error: {error}",
                ))
                continue
            wrapper = cst.MetadataWrapper(tree)
            visitor = _EmptyCatchFinder(relative(path))
            wrapper.visit(visitor)
            result.extend(visitor.findings)
        return result


RULE: HarnessCheckRule = ForbidEmptyCatchBlockRule()
